Documentation (#8)
sandilsranasinghe/wagtail-streamfield-migration-toolkit: README.md

[![License](https://img.shields.io/badge/License-BSD_3--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

[![PyPI version](https://badge.fury.io/py/streamfield-migration-toolkit.svg)](https://badge.fury.io/py/streamfield-migration-toolkit)
[![Wagtail Hallo CI](https://github.com/wagtail/streamfield-migration-toolkit/actions/workflows/test.yml/badge.svg)](https://github.com/wagtail/streamfield-migration-toolkit/actions/workflows/test.yml)

# Contents

- [Introduction](#introduction)
  - [Why data migrations?](#why-data-migrations)
- [Quick Start](#quick-start)
- [Reference](docs/REFERENCE.md)
- [Usage](docs/USAGE.md)
- [Contributing](docs/CONTRIBUTING.md)

# Introduction

This package aims to make it easier for developers using StreamField who need to write data
migrations when making changes involving blocks/block structure in the StreamField. We expose a
set of utilities for commonly made changes such as renaming or removing blocks, as well as utility
functions for recursing through existing Streamfield data and applying changes, which makes it
easier to create custom logic for applying changes too.

## Why Data Migrations?

> A streamfield is stored as a single column of JSON data in the database, where there are blocks stored (which can be nested to form complex block structures) within the JSON representation. However, as far as django is concerned when making schema migrations, everything inside this column is just a string of JSON data and the schema doesn’t change regardless of the content/structure of the StreamField since it is the same field type before and after the change. Therefore whenever changes are made to StreamFields, any existing data must be changed into the required structure by using a data migration created manually by the user.

Reviewer: Maybe just: "A StreamField is stored as a single column of JSON data in the database. Blocks are stored as structures within the JSON, and can be nested` - to avoid the slightly convoluted first sentence

While this is fairly straightforward for a very simple and small change like changing the name of a CharBlock, where you could define two functions for mapping the data and loop through all the blocks in the StreamField checking whether they are the required block type and applying the mapper accordingly; this can easily get very complicated when nested blocks and multiple fields are involved.

This package would make it easier to write data migrations for streamfield changes by providing utilities
to recurse through various streamdata structures and map changes, in addition to having several "operations"
which also handle the logic for altering the data for common use cases like renaming, removing and altering
values of blocks.

# Quick Start

## Installation

- `pip install wagtail-streamfield-migration-toolkit`
- Add `"wagtail_streamfield_migration_toolkit"` to `INSTALLED_APPS`

## Supported versions

- Python 3.7, 3.8, 3.9
- Django 3.2, 4.0
- Wagtail 4.0

## Quick Usage

> Assume we have a model `BlogPage` in app `blog` which has a streamfield `content` with a child
> streamblock named `mystream` which has a child char block named `field1` which we want to rename to

Reviewer 2: This is a little hard to follow. Perhaps we could just write the code?

```suggestion
Assume we have a model `BlogPage` in app `blog`, defined as follows:

```python
class BlogPage(Page):
    content = StreamField([("field1", blocks.CharBlock())])```

If we want to rename "field1" to "block1", we would use the following migration
```

[![PyPI version](https://badge.fury.io/py/streamfield-migration-toolkit.svg)](https://badge.fury.io/py/streamfield-migration-toolkit)
[![Wagtail Hallo CI](https://github.com/wagtail/streamfield-migration-toolkit/actions/workflows/test.yml/badge.svg)](https://github.com/wagtail/streamfield-migration-toolkit/actions/workflows/test.yml)

# Contents

- [Introduction](#introduction)
- [Quick Start](#quick-start)
- [Reference](docs/REFERENCE.md)
- [Usage](docs/USAGE.md)
- [Contributing](docs/CONTRIBUTING.md)

# Introduction

> This package aims to make it easier for developers using StreamField who need to write data 

Reviewer 2: A short description of the problem the package addresses would be good here - imagine you're a new Wagtail developer and don't know the internals well yet. In other words, _why_ does a developer need to write data migrations, and what is the cost?

functions for recursing through existing Streamfield data and applying changes, which makes it 
easier to create custom logic for applying changes too.

# Quick Start

## Installation

- `pip install streamfield-migration-toolkit`
- Add `"wagtail_streamfield_migration_toolkit"` to `INSTALLED_APPS`

## Supported versions

- Python 3.7, 3.8, 3.9
- Django 3.2, 4.0
- Wagtail 4.0

## Quick Usage

If we have a model `BlogPage` in app `blog` which has a streamfield `content` with a child block
named `field1` which we want to rename to `block1`,

```
from django.db import migrations

from wagtail_streamfield_migration_toolkit.migrate_operation import MigrateStreamData
from wagtail_streamfield_migration_toolkit.operations import RenameStreamChildrenOperation

class Migration(migrations.Migration):

    dependencies = [
        ...
    ]

    operations = [
        MigrateStreamData(
            app_name="blog",
            model_name="BlogPage",
            field_name="content",
            operations_and_block_paths=[
                (RenameStreamChildrenOperation(old_name="field1", new_name="block1"), ""),
            ]
        ),
    ]

```

> 

Reviewer: I think worth a link to the usage docs since this is pretty lightweight?
